### PR TITLE
runLaTeX: added dependency search path argument.

### PR DESCRIPTION
**Pull request by ilyakooo0**
###### Motivation for this change

It is sometimes useful to specify where LaTeX dependencies should be searched for. (For example when the root file is processed as a separate derivation.)

There was also a `TODO` mentioning the exact same feature.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/typesetting/tex/nix/default.nix [hunk 1]
```diff
@@ -63,17 +64,20 @@ rec {
...
           foundDeps = dep: xs:
             let
               exts =
                 if dep.type == "img" then [".pdf" ".png" ".ps" ".jpg"]
                 else if dep.type == "tex" then [".tex" ""]
                 else [""];
               fn = pkgs.lib.findFirst (fn: builtins.pathExists fn) null
-                (map (ext: dirOf key + ("/" + dep.name + ext)) exts);
+                (builtins.concatLists
+                  (map (path: map (ext: toString "${path}/${dep.name}${ext}") exts) searchPath')
```
Comment (veprbl) [inline pkgs/tools/typesetting/tex/nix/default.nix:79]: Why do you have `toString` here?

Also `"${path}/${file}"` is not the same as `toString (path + "/${file}")`, is that change intentional?

Comment (ilyakooo0) [reply pkgs/tools/typesetting/tex/nix/default.nix]: Removing the `toString` from line 79 seems to not break things, although I am not 100% confident that will always be the case.

removing the `toString` from line 51 definitely breaks things. (At least with the input I tried it with)

Comment (veprbl) [reply pkgs/tools/typesetting/tex/nix/default.nix]: > Is there a place I can read about what the difference is?

Sorry, didn't mean to be cryptic. The "${path}/${file}" will first put the whole "path" directory into the nix store and the resulting string will point to a file in that store directory. The `path + "/${file}"` doesn't try to reduce "path" itself to a string, the result will be still a path (in nixlang terms), so that allows for just a file to be placed into the store.

File: pkgs/tools/typesetting/tex/nix/default.nix [hunk 2]
```diff
@@ -63,17 +64,20 @@ rec {
             { rootFile = baseNameOf (toString rootFile); src = key; }
             "${pkgs.perl}/bin/perl ${./find-includes.pl}");
 
+          searchPath' = [(dirOf key)] ++ searchPath;
```
Comment (veprbl) [inline pkgs/tools/typesetting/tex/nix/default.nix:67]: This makes `dirOf key` take precedence over `searchPath`, do we want that?

Comment (ilyakooo0) [reply pkgs/tools/typesetting/tex/nix/default.nix:67]: I think there is no obviously right answer, however, what I was thinking is: if the file being imported is right next to a file (in the same folder), it seems like it should take precedence.

Comment (veprbl) [reply pkgs/tools/typesetting/tex/nix/default.nix:67]: I guess that should depend on what would be the use case for `searchPath`. Not obvious to me either.

File: pkgs/tools/typesetting/tex/nix/default.nix [hunk 3]
```diff
@@ -43,11 +44,11 @@ rec {
   # \input{}), images (e.g. \includegraphics{}), bibliographies, and
   # so on.
   findLaTeXIncludes =
-    { rootFile
+    { rootFile, searchPath
     }:
 
     builtins.genericClosure {
-      startSet = [{key = rootFile;}];
+      startSet = [{key = toString rootFile;}];
```
Comment (veprbl) [inline pkgs/tools/typesetting/tex/nix/default.nix:51]: You said, not having this breaks things. Do you understand why?

Comment (ilyakooo0) [reply pkgs/tools/typesetting/tex/nix/default.nix:51]: This was the first time I was made aware of string contexts in Nix: I started getting errors like `cannot compare a path with a string with context`.

From what I have gathered, the problem is that in the implementation of `builtins.genericClosure` the values of `key` are compared with each other, and Nix, being a strictly typed dynamic language, sees strings with a context and ordinary strings(paths) as distinct types, and throws an error. 

I have experimentally found that explicitly casting `rootFile` to a string (calling `toString`) seems to avoid the error (at least on the inputs I have). However, I have not been able to track down exactly where and why the string context appears. Because of this, I am not entirely confident that this code will work with all seemingly valid input. 

In my estimation, casting all values of `key` to strings explicitly (calling `toString`) should work in all cases since `key` will always be the same type. This is the reason I expressed my doubts in the comment: https://github.com/NixOS/nixpkgs/pull/119883#discussion_r616079867.

File: pkgs/tools/typesetting/tex/nix/default.nix (hunk 2 above)
Comment (ilyakooo0) [reply pkgs/tools/typesetting/tex/nix/default.nix:67]: For my use case, it doesn't really matter either way.

What I am doing is: creating a separate derivation which calls `lhs2TeX` with the root `.tex` file as input. I want to be able to import other local files in my resulting root `.tex`. I don't want to analyze and copy the dependencies in the `lhs2TeX` derivation as it is duplicate work and wastes space.

Aside: I create a custom derivation because I think the current `lhs2TeX` functions from nixpkgs don't work, but that is a separate problem.

**Comment by ilyakooo0** [comment]
*Turns out*, calling `toString` on a path actually converts to a string referring to the source file (not in `/nix/store`). So, this PR is very much broken. I am not planning to fix it any time soon.

The way I resolved the issue in my particular use case was by explicitly specifying all files in `extraFiles`.

**Comment by veprbl** [comment]
> _Turns out_, calling `toString` on a path actually converts to a string referring to the source file (not in `/nix/store`).

Well, you could always replace `toString` with string interpolation `${}`. I still could not understand why did you need toString in the first place and haven't had time to look into that.

**Comment by ilyakooo0** [comment]
Without the `toString` I get `cannot compare a path with a string with context` during evaluation. I didn't have to dig deeper myself.
